python/pyspark/accumulators.py

```diff
@@ -176,44 +193,44 @@ class AccumulatorParam(object):
     [7.0, 8.0, 9.0]
     """
 
-    def zero(self, value):
+    def zero(self, value: T) -> T:
         """
         Provide a "zero value" for the type, compatible in dimensions with the
         provided `value` (e.g., a zero vector)
         """
         raise NotImplementedError
 
-    def addInPlace(self, value1, value2):
+    def addInPlace(self, value1: T, value2: T) -> T:
         """
         Add two values of the accumulator's data type, returning a new value;
         for efficiency, can also update `value1` in place and return it.
         """
         raise NotImplementedError
 
 
-class AddingAccumulatorParam(AccumulatorParam):
+class AddingAccumulatorParam(AccumulatorParam[U]):
 
     """
     An AccumulatorParam that uses the + operators to add values. Designed for simple types
     such as integers, floats, and lists. Requires the zero value for the underlying type
     as a parameter.
     """
 
-    def __init__(self, zero_value):
+    def __init__(self, zero_value: U):
         self.zero_value = zero_value
 
-    def zero(self, value):
+    def zero(self, value: U) -> U:
         return self.zero_value
 
-    def addInPlace(self, value1, value2):
+    def addInPlace(self, value1: U, value2: U) -> U:
         value1 += value2
```

Comment:
`__add__` is not necessary for `__iadd__`. You can implement the latter one without the former. However, Python data model allows fallbacks from `x += y` to `x = x + y` when only `__add__` is implemented).

So if addition of `__add__` makes type checker happy, it might indicate a deeper problem, probably with the protocol

Reply:
OK, so reducing this a bit:

```python
# main.py
 
from __future__ import annotations
from typing import Generic, Protocol, TypeVar     
 
 
T = TypeVar("T")
U = TypeVar("U", bound=SupportsIAdd)

 
class SupportsIAdd(Protocol):
    def __iadd__(self: T, other: T) -> T: ...        

 
class AddingAccumulatorParam(Generic[U]):
    def addInPlace(self, value1: U, value2: U) -> U:
        value1 += value2
        return value1

```

```
main.py:17: error: Unsupported left operand type for + ("U")
Found 1 error in 1 file (checked 1 source file)
```

but after replacing `__iadd__` with `__add__` passes 

```python
# main.py
 
from __future__ import annotations
from typing import Generic, Protocol, TypeVar     
 
 
T = TypeVar("T")
U = TypeVar("U", bound=SupportsIAdd)

 
class SupportsIAdd(Protocol):
    def __add__(self: T, other: T) -> T: ...         

 
class AddingAccumulatorParam(Generic[U]):
    def addInPlace(self, value1: U, value2: U) -> U:
        value1 += value2
        return value1
```

I honestly don't see why, but there are all kinds of subtleties around `__iadd__`, `__radd__` and `__add__`. Anyone?

Reply:
It seems that the problem is somehow related to combination of `__iadd__` and `TypeVar`. Further reducing the example ‒ this still fails 

```python
# main.py

from __future__ import annotations
from typing import Any, TypeVar              


U = TypeVar("U", bound="SupportsIAdd")        


class SupportsIAdd:          
    def __iadd__(self: U, other: Any) -> U: ...                 


class AddingAccumulatorParam:            
    def addInPlace(self, value1: U, value2: U) -> U: 
        value1 += value2
        return value1
````

but substituting `U` with `SupportsIAdd`:


```python
# main.py

from __future__ import annotations
from typing import Any, TypeVar              


U = TypeVar("U", bound="SupportsIAdd")        


class SupportsIAdd:          
    def __iadd__(self, other: Any) -> SupportsIAdd: ...         


class AddingAccumulatorParam:            
    def addInPlace(self, value1: SupportsIAdd, value2: SupportsIAdd) -> SupportsIAdd:            
        value1 += value2
        return value1

```

makes mypy happy. It don't think it helps us to resolve it here, but narrows down the scope, in  case we want to escalate this to mypy folks.

Reply:
Posted on mypy gitter: https://gitter.im/python/typing?at=61b87c2f197fa95a1cc2c2a4

Reply:
And corresponding mypy issue https://github.com/python/mypy/issues/11745